scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ServerHandshaker.java

```diff
@@ -521,13 +528,17 @@ private void createServerKeyExchange(final ClientHello clientHello, final DTLSFl
 
 	private boolean createCertificateRequest(final ClientHello clientHello, final DTLSFlight flight) throws HandshakeException {
 
-		if ((clientAuthenticationWanted || clientAuthenticationRequired) && signatureAndHashAlgorithm != null) {
+		if ((clientAuthenticationWanted || clientAuthenticationRequired) && negotiatedSignatureAndHashAlgorithms != null
+				&& !negotiatedSignatureAndHashAlgorithms.isEmpty()) {
```

Comment:
`negotiatedSignatureAndHashAlgorithms != null && !negotiatedSignatureAndHashAlgorithms.isEmpty()`

**I probably didn't get the point here**, but this sounds a bit strange that if `negotiatedSignatureAndHashAlgorithms` or `negotiatedSignatureAndHashAlgorithms` are `null` we just don't create a certificate request.

This sounds more critical to me.

Reply:
If there are no common sig_hash_algo, the client will not  be able to send successfully it's certificate, though [isSignedWithSupportedAlgorithm](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/CertificateRequest.java#L499) will fail.

Reply:
Before this approach, it is "assumed/required", that both sides support the {sha256,ecdsa}. And though all californium certificates use that, all work well. But it comes also with some inflexibility. The implementation was mainly triggered by PR #1253 , though I prefer to support functionality on both sides.

Reply:
I'm totally agree to integrate the support of this extention :+1: 

On this point : 
 -  either `negotiatedSignatureAndHashAlgorithms` can not be null or empty because check was done before (and so this check is useless)
 - or if it is null or empty and null or empty means the is no  "compatible" sign/hash then we must stop the handshake instead of just "not sending the CertificateRequest"

Maybe this is the current behavior but when you only read this method, it doesn't look like.

Reply:
In case of a configuration for both PSK and EC cipher suites, the negotiated list maybe empty. Though the `createCertificateRequest` is always called, also for PSK, it checks itself, if the request should be included or not.

Reply:
just to be clear for PSK this does make sense to send a certificateRequest ?

Reply:
Though PSK requires the same secret on both sides and with that, no additional data is needed, PSK doesn't use a certificate request.

Reply:
Just additionally: 
The signing (+ hash) itself is used for the certificate chain and the "server key exchange" and the (client) "certificate verify" message. It's send in the "certificate request" to ensure, that the clients certificate and "certificate verify" also matches the server's limitations.

Reply:
> Though the createCertificateRequest is always called

So it seems to me not so intuitive, I think we should not call this method for PSK.

Now, imagine that PSK is excluded. What happens if `negotiatedSignatureAndHashAlgorithms` is null or empty. It seems to me that this should stop the handshake and here this looks like it just imply that we will not send the CertificateRequest.

Reply:
```
private void negotiateCipherSuite(final ClientHello clientHello, final HelloExtensions serverHelloExtensions) throws HandshakeException {
	...
	// if none of the client's proposed cipher suites matches throw exception
	AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.HANDSHAKE_FAILURE, session.getPeer());
	throw new HandshakeException("Client proposed unsupported cipher suites only", alert);
}
```

If there is no matching cipher suite, the handshake is aborted here.

```
if (cipherSuite.requiresServerCertificateMessage()) {
	SignatureAndHashAlgorithm algo = getNegotiatedSignatureAndHashAlgorithm(signatures);
	if (algo != null) {
		negotiatedSignatureAndHashAlgorithms = signatures;
		session.setSignatureAndHashAlgorithm(algo);
	}
}
```

That check ensures, that `negotiatedSignatureAndHashAlgorithms` indicates, if ECDSA cipher suite is selected and therefore the signature and hash algorithms are negotiated.

For me you seems to have different idea's. I would prefer to add more docu to the current implementation.